pkg/objstore/cos/cos.go

```diff
@@ -0,0 +1,423 @@
+package cos
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"math/rand"
+	"net/http"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/improbable-eng/thanos/pkg/objstore"
+	"github.com/improbable-eng/thanos/pkg/runutil"
+	"github.com/mozillazg/go-cos"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	yaml "gopkg.in/yaml.v2"
+)
+
+const (
+	opObjectsList  = "ListBucket"
+	opObjectInsert = "PutObject"
+	opObjectGet    = "GetObject"
+	opObjectHead   = "HeadObject"
+	opObjectDelete = "DeleteObject"
+)
+
+// DirDelim is the delimiter used to model a directory structure in an object store bucket.
+const dirDelim = "/"
+
+// Bucket implements the store.Bucket interface against cos-compatible(Tencent Object Storage) APIs.
+type Bucket struct {
+	logger   log.Logger
+	client   *cos.Client
+	opsTotal *prometheus.CounterVec
+	name     string
+}
+
+// cosConfig encapsulates the necessary config values to instantiate an cos client.
+type cosConfig struct {
+	Bucket    string `yaml:"bucket"`
+	Region    string `yaml:"region"`
+	AppId     string `yaml:"appid"`
+	SecretKey string `yaml:"secret-key"`
+	SecretId  string `yaml:"secret-id"`
+}
+
+// Validate checks to see if mandatory cos config options are set.
+func Validate(conf cosConfig) error {
+	if conf.Bucket == "" ||
+		conf.AppId == "" ||
+		conf.Region == "" ||
+		conf.SecretId == "" ||
+		conf.SecretKey == "" {
+		return errors.New("insufficient cos configuration information")
+	}
+	return nil
+}
+
+func NewBucket(logger log.Logger, conf []byte, reg prometheus.Registerer, component string) (*Bucket, error) {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+
+	var config cosConfig
+	if err := yaml.Unmarshal(conf, &config); err != nil {
+		return nil, errors.Wrap(err, "parsing cos configuration")
+	}
+	if err := Validate(config); err != nil {
+		return nil, errors.Wrap(err, "validate cos configuration")
+	}
+
+	bucketUrl := cos.NewBucketURL(config.Bucket, config.AppId, config.Region, true)
+
+	b, err := cos.NewBaseURL(bucketUrl.String())
+	if err != nil {
+		return nil, errors.Wrap(err, "initialize cos base url")
+	}
+
+	client := cos.NewClient(b, &http.Client{
+		Transport: &cos.AuthorizationTransport{
+			SecretID:  config.SecretId,
+			SecretKey: config.SecretKey,
+		},
+	})
+
+	bkt := &Bucket{
+		logger: logger,
+		client: client,
+		opsTotal: prometheus.NewCounterVec(prometheus.CounterOpts{
```

Comment:
do we need this? If we are using `BucketWithMetrics` anyway?

Reply:
It seems to be not necessary at this moment because of don't support multiple stores in each store components. I will create another PR to remove all the metrics for all object stores.